### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Galois): decomposition in connected components

### PR DESCRIPTION
**Pull request by chrisflav**
Shows that any object in a Galois category is the sum of its connected components. Also shows that the connected components are unique up to isomorphism.

---

- [x] depends on: #10400
- [x] depends on: #10418

This will be used in the followup PR to show that the fiber of any object is represented by a Galois object, which essentially shows the pro-representability of fiber functors.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Galois/Basic.lean [hunk 1]
```diff
@@ -178,6 +180,14 @@ lemma not_initial_iff_fiber_nonempty (X : C) : (IsInitial X → False) ↔ Nonem
 lemma not_initial_of_inhabited {X : C} (x : F.obj X) (h : IsInitial X) : False :=
   ((initial_iff_fiber_empty F X).mp ⟨h⟩).false x
 
+/-- An object that is neither initial or connected has a non-trivial subobject. -/
+lemma has_non_trivial_subobject_of_not_isConnected_of_not_initial (X : C) (hc : ¬ IsConnected X)
+    (hi : IsInitial X → False) :
+    ∃ (Y : C) (v : Y ⟶ X), (IsInitial Y → False) ∧ Mono v ∧ (¬ IsIso v) := by
+  by_contra h2
+  simp only [exists_and_left, not_exists, not_and, not_not] at h2
+  exact hc ⟨hi, fun Y i hm hni ↦ h2 Y hni i hm⟩
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Basic.lean:189]: Does this work
```suggestion
  contrapose! hc
  exact ⟨hi, fun Y i hm hni ↦ hc Y hni i hm⟩
```

**Review by jcommelin — COMMENTED** [review]
Small comment. I'll review the big file later.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+namespace Cofan
+
+variable {ι₁ ι₂ : Type} {X X₁ X₂ : C} {f₁ : ι₁ → C} {f₂ : ι₂ → C}
+    (g₁ : (i : ι₁) → (f₁ i) ⟶ X₁) (g₂ : (i : ι₂) → (f₂ i) ⟶ X₂) (v₁ : X₁ ⟶ X) (v₂ : X₂ ⟶ X)
+    (h₁ : IsColimit (Cofan.mk X₁ g₁)) (h₂ : IsColimit (Cofan.mk X₂ g₂))
+    (h : IsColimit (BinaryCofan.mk v₁ v₂))
+
+/-- For maps `f₁ : ι₁ → C`, `f₂ : ι₂ → C`, families of morphisms `g₁ i : f₁ i ⟶ X₁`,
+`g₂ i : f₂ i ⟶ X₂` and morphisms `v₁ : X₁ ⟶ X`, `v₂ : X₂ ⟶ X`, construct one family of
+morphisms on `ι₁ ⊕ ι₂` -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:43]:
```suggestion
morphisms indexed by `ι₁ ⊕ ι₂` -/
```

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+universe u₁ u₂ w
+
+namespace CategoryTheory
+
+open Limits Functor
+
+variable {C : Type u₁} [Category.{u₂} C]
+
+namespace Cofan
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: The two declarations you've put in this namespace probably belong in another file.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    obtain ⟨hi⟩ := nhi
+    exact has_decomp_connected_components_aux_initial X hi
+
+/-- In a Galois category, every object is the sum of connected objects. -/
+theorem has_decomp_connected_components (X : C) :
+    ∃ (ι : Type) (f : ι → C)
+    (g : (i : ι) → f i ⟶ X)
+    (_ : IsColimit (Cofan.mk X g)),
+    (∀ i, IsConnected (f i)) ∧ Finite ι := by
+  obtain ⟨F, ⟨hf⟩⟩ := @GaloisCategory.hasFiberFunctor C _ _
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:147]: To avoid repetition of this sort of `obtain` in the future, could you put a definition `GaloisCategory.getFiberFunctor` (which gives the `F`) and an instance `FiberFunctor GaloisCategory.getFiberFunctor`, in the `Basic` file?

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  exact has_decomp_connected_components_aux F (Nat.card <| F.obj X) X rfl
+
+/-- In a Galois category, every object is the sum of connected objects. -/
+theorem has_decomp_connected_components' (X : C) :
+    ∃ (ι : Type) (_ : Finite ι) (f : ι → C) (_ : ∐ f ≅ X), ∀ i, IsConnected (f i) := by
+  obtain ⟨ι, f, g, hl, hc, hf⟩ := has_decomp_connected_components X
+  refine ⟨ι, hf, f, colimit.isoColimitCocone ⟨Cofan.mk X g, hl⟩, hc⟩
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:154]:
```suggestion
  exact ⟨ι, hf, f, colimit.isoColimitCocone ⟨Cofan.mk X g, hl⟩, hc⟩
```

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  obtain ⟨ι, f, g, hl, hc, he⟩ := has_decomp_connected_components X
+  have : Fintype ι := Fintype.ofFinite ι
+  let s : Cocone (Discrete.functor f ⋙ F) := F.mapCocone (Cofan.mk X g)
+  let s' : IsColimit s := isColimitOfPreserves F hl
+  have : ∃ (j : Discrete ι) (z : (Discrete.functor f ⋙ F).obj j), s.ι.app j z = x :=
+    Concrete.isColimit_exists_rep _ s' x
+  obtain ⟨⟨j⟩, z, h⟩ := this
+  refine ⟨f j, g j, z, ⟨?_, hc j, MonoCoprod.mono_inj _ (Cofan.mk X g) hl j⟩⟩
+  subst h
+  simp only [mapCocone_pt, Cofan.mk_pt, mapCocone_ι_app, Discrete.functor_obj, Cofan.mk_ι_app]
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:170]:
```suggestion
  obtain ⟨⟨j⟩, z, h⟩ := Concrete.isColimit_exists_rep _ s' x
  refine ⟨f j, g j, z, ⟨?_, hc j, MonoCoprod.mono_inj _ (Cofan.mk X g) hl j⟩⟩
  subst h
  rfl
```

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/Galois/Decomposition.lean (hunk 3 above)
Comment (dagurtomas) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: I don't know if it should be a new file or an existing one, but it's too general for this file.

Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: I agree, I don't know where to put them though (the reason why I ultimately left them here).

File: Mathlib/CategoryTheory/Galois/Basic.lean (hunk 1 above)
Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Basic.lean:189]: Indeed, this works! I did not know the `!` option, that is quite neat.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean (hunk 3 above)
Comment (dagurtomas) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: Did you try #find_home?

Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: Yes, but the result is this file :/ I had some discussions on this with @joelriou here: https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Combine.20coproducts

Ideally one could build a suitably general API for combining cofans, which in theory already exists with the `GradedObject` API. In practice though, using this API amounts to showing some cocone categories are isomorphic in every applied situation which turns out to be longer than just reproving the colimit property (as its done here).

Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: I would be fine with creating a new file `CombinedProducts.lean` in the `Limits.Shapes` folder and to collect specialized helpers like this one there. We could also put the one posted in the linked Zulip chat above for Sigma types.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+open Limits Functor
+
+variable {C : Type u₁} [Category.{u₂} C]
+
+namespace Cofan
+
+variable {ι₁ ι₂ : Type} {X X₁ X₂ : C} {f₁ : ι₁ → C} {f₂ : ι₂ → C}
+    (g₁ : (i : ι₁) → (f₁ i) ⟶ X₁) (g₂ : (i : ι₂) → (f₂ i) ⟶ X₂) (v₁ : X₁ ⟶ X) (v₂ : X₂ ⟶ X)
+    (h₁ : IsColimit (Cofan.mk X₁ g₁)) (h₂ : IsColimit (Cofan.mk X₂ g₂))
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:38]: Instead of using variables `X₁`, `g₁`, etc, it would be better to have a variable `c₁ : Cofan f₁` and `h₁ : IsColimit c₁`. A general `BinaryCofan` instead of a `BinaryCofan.mk` would also be better.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean:34]: "When in doubt, create a new file"

(I just made this up, but it sounds true :grinning:)

Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean]: I created a `CombinedProducts.lean` and added the version for `Fan`s. In a followup PR I can add other similar constructors floating around in the mentioned Zulip stream.

Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean]: One might want to refactor the file structure to move `Products.lean` to `Products/Basic.lean` and move the `CombinedProducts.lean` into that folder. If this is wanted, I can do it, but suggest do it after this PR is merged, to not obfuscate the diff.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by chrisflav** [comment]
Thanks for all the reviews!

**Comment by chrisflav** [comment]
I don't know why the build keeps failing, I just merged master, maybe that helps. Could someone send this to bors again?

**Comment by sgouezel** [comment]
Let me try again
bors r+
